### PR TITLE
60 - History sync from pcs4 consumer

### PR DESCRIPTION
**Pull request by janid1967**
https://github.com/equinor/cc-toolbox-tasks/issues/60

### CODE REVIEW

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+        var busEvent = context.Message;
+
+        ValidateMessage(busEvent);
+
+        if (!await historyItemRepository.ExistsAsync(busEvent.ProCoSysGuid, context.CancellationToken))
+        {
+            var document = CreateHistoryItemEntity(busEvent);
+            historyItemRepository.Add(document);
+            await unitOfWork.SaveChangesAsync(context.CancellationToken);
+            logger.LogInformation($"{nameof(PunchItemChangeHistoryEvent)} Message Consumed: {{MessageId}} \n Guid {{Guid}}",
```
Comment (DownGoat) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs:29]: Avoid string concatenation for log statements

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+public record PunchItemChangeHistoryEvent(
+    Guid ProCoSysGuid,
+    Guid PunchGuid,
+    string FieldName,
+    string? OldValue,
+    string? OldValueLong,
+    string? NewValue,
+    string? NewValueLong,
+    string ChangedBy,
+    DateTime ChangedAt
```
Comment (TordJoranger) [inline src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs:165]: ChangedAt is nullable

Comment (TordJoranger) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs:165]: ChangedBy also

File: src/Equinor.ProCoSys.Completion.Infrastructure/EntityConfigurations/HistoryItemConfiguration.cs [hunk 3]
```diff
@@ -36,5 +36,9 @@ public void Configure(EntityTypeBuilder<HistoryItem> builder)
             .OnDelete(DeleteBehavior.NoAction);
 
         // todo History: Create index
+
+        builder
```
Comment (DownGoat) [inline src/Equinor.ProCoSys.Completion.Infrastructure/EntityConfigurations/HistoryItemConfiguration.cs:40]: Now the entity has two keys Id, and Guid? Maybe we should remove the Id column?

**Review by DownGoat — APPROVED** [review]

**Review by TordJoranger — APPROVED** [review]

File: src/Equinor.ProCoSys.Completion.Infrastructure/EntityConfigurations/HistoryItemConfiguration.cs (hunk 3 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.Infrastructure/EntityConfigurations/HistoryItemConfiguration.cs:40]: we tried last week with other entities, but the rewrite job got to big for now. We can create an issue for it and look into it again later

File: src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs (hunk 2 above)
Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs:165]: No such records with null for ChangedAt / By in sandbox. Did you check in prod @TordJoranger ?

Comment (janid1967) [reply src/Equinor.ProCoSys.Completion.WebApi/Synchronization/PunchItemChangeHistoryEventConsumer.cs]: image

**Pull request merged**